### PR TITLE
Fix #3906: Incorrect ouput for rich text in hints and solution [Blocked: #3852]

### LINKED ISSUE #3906: [A11Y] Incorrect output for rich-text in Hints and Solution
Incorrect output for rich-text in Hints and Solution.

### PR DESCRIPTION
**Pull request by vrajdesai78**
## Explanation
Fix #3906:  Parsed HTML to normal text while setting Content Description for Hints and Solution. 

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

https://user-images.githubusercontent.com/43074241/155514660-2daadc00-a73d-4d51-a702-b52ed341d09a.mp4

### CODE REVIEW

**Comment by vrajdesai78** [comment]
@BenHenning I figured out that there were only a few cases (I am able to find only 1) where HTML is not parsing properly for content description before changing this code. How can I approach to write a test case for this issue?

File: app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt]: We should only be using our own HtmlParser when parsing HTML to ensure consistency and support for custom tags.

I think this will require doing something similar to the parseOppiaHtml call in HintsAndSolutionDialogFragmentPresenter, though you'll need to update parseOppiaHtml itself to support not taking a TextView (for cases when we just want to parse into spannables without tweaking the textview itself).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Apologies for the delayed review. Left one comment--PTAL.

**Comment by BenHenning** [comment]
> @BenHenning I figured out that there were only a few cases (I am able to find only 1) where HTML is not parsing properly for content description before changing this code. How can I approach to write a test case for this issue?

@vrajdesai78 you would need to write a test in the suite corresponding to the view model (it may actually be a test suite for a fragment or activity that indirectly uses the view model), load the content that fails, and verify that it now has the correct content description. You should verify that the test also fails without the fix in place to make sure that it's correctly passing when it passes with the fix.

**Comment by rt4914** [comment]
@vrajdesai78 As per our discussion earlier the most important thing in this PR is writing test which I am not sure about. So once you have written it as per Ben's comment please let me know and I will do a pass on it.

File: app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt
Comment (vrajdesai78) [reply app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt]: I have updated parseOppiaHtml and related code.

**Comment by vrajdesai78** [comment]
> > @BenHenning I figured out that there were only a few cases (I am able to find only 1) where HTML is not parsing properly for content description before changing this code. How can I approach to write a test case for this issue?
> 
> @vrajdesai78 you would need to write a test in the suite corresponding to the view model (it may actually be a test suite for a fragment or activity that indirectly uses the view model), load the content that fails, and verify that it now has the correct content description. You should verify that the test also fails without the fix in place to make sure that it's correctly passing when it passes with the fix.

I have added a test case `testStateFragment_showHintsAndSolution_hasCorrectContentDescription` in StateFragmentTest. In this test I am checking that hint has correct content description.

Test case failing if HTML is not parsed for content description (Both espresso and robolectric).
Screenshot from 2022-02-26 21-09-58

Test case passing after parsing HTML properly (On both espresso and robolectric).
Screenshot from 2022-02-27 00-12-33

@BenHenning PTAL.

File: app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt:36]: I think this changes based on whether an exploration or question is being loaded. We probably need to pass in the entity type from the fragment (which I expect we must do for the content itself since hints can contain images).

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: This test probably should be in StateFragmentLocalTest (we generally don't test hints here because we sometimes would need to wait seconds for them to show up).

File: utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt]: Could you consolidate this and the above into one call, instead?

File: utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt]: I think it's fine to have a nullable property for a helper, but for the public API I suggest making things more explicit by having 1 version that takes a non-null image retriever and another which takes nothing but the HTML.

File: utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt]: Ditto here--we ought to have two versions of the function to keep things clearer. I suggest using a private helper to consolidate the two (and that can certainly use a nullable parameter to make things simpler).

**Review by BenHenning — COMMENTED** [review]
Thanks @vrajdesai78. This is definitely on the right track. Besides the comments I left, please also:
- Add tests for the content handler and html parser changes
- Make sure all CI checks are passing (a bunch are currently failing)

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (vrajdesai78) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: @BenHenning Done

File: utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt
Comment (vrajdesai78) [reply utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt]: @BenHenning Done

File: utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt
Comment (vrajdesai78) [reply utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt]: Done

File: utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt
Comment (vrajdesai78) [reply utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt]: I have created another version of parseOppiaHtml which takes only rawString as a parameter, but I didn't how can I use a private helper.

File: app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt
Comment (vrajdesai78) [reply app/src/main/java/org/oppia/android/app/hintsandsolution/HintsViewModel.kt:36]: I am not able to understand that how can I use entity type from fragment. Can you elaborate a little bit. Thanks @BenHenning

**Comment by vrajdesai78** [comment]
@BenHenning can you give me a hint that why CI checks are failing and how can I debug it. Thanks

File: app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt
Comment (BenHenning) [inline app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:1326]: Nit: this should be directly imported (like other Espresso functions).

File: app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt
Comment (BenHenning) [inline app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:1323]:
```suggestion
      // Ensure the hint description is correct and doesn't contain any HTML.
```

Fixes capitalization & punctuation, but it also rewords to be a bit clearer.

File: utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt:214]: This should still be using CustomContentHandler.

It might actually be worth waiting until #3852 is merged since it's introducing a very similar function with tests which might make this a lot easier and avoid redundant work.

**Review by BenHenning — COMMENTED** [review]
Thanks @vrajdesai78. After looking at this again, I think it might actually be worth waiting until #3852 is finished since it's introducing the exact HtmlParser functionality that you need (i.e. parsing HTML without including an image provider).

**Comment by BenHenning** [comment]
> @BenHenning can you give me a hint that why CI checks are failing and how can I debug it. Thanks

@vrajdesai78 these are Bazel failures that are coming up because you're changing a view model to depend on HtmlParser, and the view_models library doesn't include the needed HtmlParser dependency in its deps list:

https://github.com/oppia/oppia-android/blob/2004db709549619cca811142b627654cc3e29a92/app/BUILD.bazel#L632

The deps list will need to be updated to include it.

File: utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt
Comment (vrajdesai78) [reply utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt:214]: Ok, then I will wait for #3852 to get merged and after that I will do required changes.

**Comment by vrajdesai78** [comment]
Keep this PR open

**Comment by vrajdesai78** [comment]
Keep this PR open

**Pull request closed without merging**
